core/lib/Cardano/Address/Script.hs

```diff
@@ -200,6 +228,62 @@ keyHashFromBytes bytes
     | BS.length bytes /= hashSize = Nothing
     | otherwise = Just $ KeyHash bytes
 
+data ErrorKeyHashFromText =
+      ErrorKeyHashFromTextInvalidString (AbstractEncoding ())
+    | ErrorKeyHashFromTextWrongEncoding
+    | ErrorKeyHashFromTextWrongPayload
+    | ErrorKeyHashFromTextWrongHrp
+    | ErrorKeyHashFromTextWrongDataPart
+    deriving (Show, Eq)
+
+showErr :: ErrorKeyHashFromText -> String
+showErr (ErrorKeyHashFromTextInvalidString EBase16) =
+    "Invalid Base16-encoded string."
+showErr (ErrorKeyHashFromTextInvalidString EBech32{}) =
+    "Invalid Bech32-encoded string."
+showErr (ErrorKeyHashFromTextInvalidString EBase58) =
+    "Invalid Base58-encoded string."
+showErr ErrorKeyHashFromTextWrongEncoding =
+    "Verification key hash must be must be encoded as \
+    \base16, bech32 or base58."
+showErr ErrorKeyHashFromTextWrongPayload =
+    "Verification key hash must contain exactly 28 bytes."
+showErr ErrorKeyHashFromTextWrongHrp =
+    "Verification key hash must have 'script_vkh' hrp when Bech32-encoded."
+showErr ErrorKeyHashFromTextWrongDataPart =
+    "Verification key hash is Bech32-encoded but has wrong data part."
+
+-- | Construct a 'KeyHash' from 'Text'. Either hex encoded text or
+-- Bech32 encoded text with `script_vkh` hrp is expected. Also
+-- binary payload is expected to be composed of 28 bytes.
+--
+-- @since 3.0.0
+keyHashFromText :: Text -> Either ErrorKeyHashFromText KeyHash
+keyHashFromText txt =  case detectEncoding str of
+        Just EBase16 -> case fromBase16 (toBytes str) of
+            Left _ -> Left $ ErrorKeyHashFromTextInvalidString EBase16
+            Right bytes -> checkPayload bytes
+        Just EBech32{} -> fromBech32
+        Just EBase58 -> case fromBase58 (toBytes str) of
```

Comment:
I don't think we should support base58 for key hashes because it will just add confusion. I would even be mildly in favour of only using bech32 and not even supporting hex. The thing I like about bech32 is the HRP - which base16 and base58 don't have. On the other hand, supporting hex is good for interoperability.

Reply:
now support only to bech32 and base16 (I think some people are very attached to this format - like me;-))  -> https://github.com/input-output-hk/cardano-addresses/pull/84/commits/6a1ebd16d433c16b5d8919743ac9a3873248e612